### PR TITLE
ECC support for icatoken and common code

### PR DESCRIPTION
**Pull request by jschmidb**

### CODE REVIEW

File: usr/lib/pkcs11/cca_stdll/cca_specific.c [hunk 1]
```diff
@@ -3328,6 +3667,36 @@ CK_RV token_specific_object_add(STDLL_TokData_t *tokdata, SESSION *sess, OBJECT
 			   " imported\n", attr->ulValueLen);
 		break;
 
+    case CKK_EC:
+        rc = template_attribute_find(object->template, CKA_CLASS, &attr);
+        if (rc == FALSE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        } else
```
Comment (dodys) [inline usr/lib/pkcs11/cca_stdll/cca_specific.c]: Please add parenthesis to else

File: testcases/crypto/ec_func.c [hunk 2]
```diff
@@ -162,6 +162,566 @@ _signVerifyParam signVerifyInput[] = {
...
+    if (!mech_supported(SLOT_ID, CKM_EC_KEY_PAIR_GEN)) {
+        testcase_skip("Slot %u doesn't support CKM_EC_KEY_PAIR_GEN\n", (unsigned int) SLOT_ID);
+        goto testcase_cleanup;
+    }
+    if (!mech_supported(SLOT_ID, CKM_ECDH1_DERIVE)) {
+        testcase_skip("Slot %u doesn't support CKM_ECDH1_DERIVE\n", (unsigned int) SLOT_ID);
+        goto testcase_cleanup;
+    }
+
+    for (int i=0; i<NUMEC; i++) {
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: Do not declare int i inside the for, please do it outside. This is causing Travis CI build to fail.

File: testcases/crypto/ec_func.c [hunk 3]
```diff
@@ -162,6 +162,566 @@ _signVerifyParam signVerifyInput[] = {
...
+
+        // Check if the key lengths are equal
+        if (extr1_tmpl->ulValueLen != extr2_tmpl->ulValueLen) {
+            testcase_error("Length of public key A not equal to length of public key B");
+            goto testcase_cleanup;
+        }
+
+        // Testcase #2 - Now derive the secrets...
+
+        for (int j=0; j < NUM_KDFS; j++) {
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: Do not declare int j inside the for, please do it outside. This is causing Travis CI build to fail.

File: testcases/crypto/ec_func.c [hunk 4]
```diff
@@ -162,6 +162,566 @@ _signVerifyParam signVerifyInput[] = {
...
+
+            for (int k=0; k<NUM_SECRET_KEY_LENGTHS; k++) {
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: Do not declare int k inside the for, please do it outside. This is causing Travis CI build to fail.

File: testcases/crypto/ec_func.c [hunk 5]
```diff
@@ -162,6 +162,566 @@ _signVerifyParam signVerifyInput[] = {
...
+                CK_ULONG secretB_tmpl_len = sizeof(secretB_tmpl)/sizeof(CK_ATTRIBUTE);
+
+                CK_ATTRIBUTE  derive_tmpl[] = {
+                    {CKA_CLASS, &class, sizeof(class)},
+                    {CKA_KEY_TYPE, &key_type, sizeof(key_type)},
+                    {CKA_VALUE_LEN, &(secret_key_len[k]), sizeof(CK_ULONG)},
+                };
+                CK_ULONG secret_tmpl_len = sizeof(derive_tmpl)/sizeof(CK_ATTRIBUTE);
+
+                for (int m=0; m<NUM_SHARED_DATA; m++) {
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: Do not declare int m inside the for, please do it outside. This is causing Travis CI build to fail.

File: testcases/crypto/ec_func.c [hunk 6]
```diff
@@ -162,6 +162,566 @@ _signVerifyParam signVerifyInput[] = {
...
+        testcase_skip("Slot %u doesn't support CKM_ECDH1_DERIVE\n", (unsigned int) SLOT_ID);
+        goto testcase_cleanup;
+    }
+
+    if (is_ep11_token(SLOT_ID)) {
+        testcase_skip("This testcase uses KDFs, currently not supported by ep11token.\n");
+        goto testcase_cleanup;
+    }
+
+    for (int i=0; i<ECDH_TV_NUM; i++) {
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: Do not declare int i inside the for, please do it outside. This is causing Travis CI build to fail.

File: testcases/crypto/ec.h [hunk 7]
```diff
@@ -485,3 +485,558 @@ struct EC_TEST_VECTOR ec_tv[] = {
...
+                     0xA5,0x12,0x3C,0x61,0xAC,0xAB,0x57,0xB5,0x39,0x3D,0xCE,0x47,0x60,0x81,0x72,0xA0,
+                     0x95,0xAA,0x85,0xA3,0x0F,0xE1,0xC2,0x95,0x2C,0x67,0x71,0xD9,0x37,0xBA,0x97,0x77,
+                     0xF5,0x95,0x7B,0x26,0x39,0xBA,0xB0,0x72,0x46,0x2F,0x68,0xC2,0x7A,0x57,0x38,0x2D,
+                     0x4A,0x52,},
+        .pubkeyA =  {0x02,0x00,0x15,0x41,0x7E,0x84,0xDB,0xF2,0x8C,0x0A,0xD3,0xC2,0x78,0x71,0x33,0x49,
+                     0xDC,0x7D,0xF1,0x53,0xC8,0x97,0xA1,0x89,0x1B,0xD9,0x8B,0xAB,0x43,0x57,0xC9,0xEC,
+                     0xBE,0xE1,0xE3,0xBF,0x42,0xE0,0x0B,0x8E,0x38,0x0A,0xEA,0xE5,0x7C,0x2D,0x10,0x75,
+                     0x64,0x94,0x18,0x85,0x94,0x2A,0xF5,0xA7,0xF4,0x60,0x17,0x23,0xC4,0x19,0x5D,0x17,
+                     0x6C,0xED,0x3E,
+                     /*0x01,0x7C,0xAE,0x20,0xB6,0x64,0x1D,0x2E,0xEB,0x69,0x57,0x86,0xD8,
```
Comment (dodys) [inline testcases/crypto/ec.h]: Do we need this comment? 
Same below in pubkey8.

File: usr/lib/pkcs11/ica_s390_stdll/tok_struct.h [hunk 8]
```diff
@@ -78,9 +78,10 @@ token_spec_t token_specific = {
 	&token_specific_rsa_pss_verify,
 	&token_specific_rsa_generate_keypair,
 	// Elliptic Curve
-	NULL,			// ec_sign
-	NULL,			// ec_verify
-	NULL,			// ec_generate_keypaiir
+    NULL,
+    NULL,
+    NULL,
+    NULL,
```
[the remark this answers is not in the document]
Comment (jschmidb) [reply usr/lib/pkcs11/ica_s390_stdll/tok_struct.h]: These NULLs are replaced with the actual functions in one of the next commits, so I would prefer to leave it as is.

**Comment by jschmidb** [comment]
All items are fixed now via interactive rebase and pushed again.

File: usr/lib/pkcs11/common/mech_ec.c [hunk 9]
```diff
@@ -675,3 +691,426 @@ ec_hash_verify_final( STDLL_TokData_t      * tokdata,
...
+
+    return CKR_OK;
+}
+
+CK_RV
+pkcs_get_keytype(CK_ATTRIBUTE *attrs, CK_ULONG attrs_len,
+        CK_MECHANISM_PTR mech, CK_ULONG *type, CK_ULONG *class)
+{
+    int i;
+    CK_RV rc = CKR_TEMPLATE_INCONSISTENT;
```
Comment (dodys) [inline usr/lib/pkcs11/common/mech_ec.c]: This CKR_TEMPLATE_INCONSISTENT doesn't make sense to me.
Can't we just do:
CK_RV rc = CKR_OK; 
and then remove line 946?

File: testcases/crypto/ec_func.c [hunk 10]
```diff
@@ -725,11 +1309,16 @@ run_TransferECCKeyPairSignVerify()
...
-	funcs->C_DestroyObject(session, priv_key);
-	funcs->C_DestroyObject(session, secret_key);
-	funcs->C_DestroyObject(session, unwrapped_key);
+    if (publ_key != CK_INVALID_HANDLE)
+        funcs->C_DestroyObject(session, publ_key);
+    if (priv_key != CK_INVALID_HANDLE)
+        funcs->C_DestroyObject(session, priv_key);
+    if (secret_key != CK_INVALID_HANDLE)
+        funcs->C_DestroyObject(session, secret_key);
+    if (unwrapped_key != CK_INVALID_HANDLE)
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: Can you also align the if below just like this one?
During build, compiler is complaining about misalignment on this code.

File: usr/lib/pkcs11/common/mech_ec.c [hunk 11]
```diff
@@ -675,3 +691,426 @@ ec_hash_verify_final( STDLL_TokData_t      * tokdata,
...
+        return CKR_OK;
+    default:
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_NOT_SUPPORTED));
+        return CKR_FUNCTION_NOT_SUPPORTED;
+    }
+
+    //digest_mech.ulParameterLen = 0;   FIXME: wird durch obiges memset erledigt.
+    //digest_mech.pParameter = NULL;
+    rc = digest_mgr_init(tokdata, sess, &ctx, &digest_mech);
+    if (rc != 0) {
```
Comment (dodys) [inline usr/lib/pkcs11/common/mech_ec.c]: Please, compare rc to CKR_OK. This is the pkcs11 way.
Same on line 745.

**Comment by jschmidb** [comment]
mech_ec.c pushed again with a bug fix.

**Comment by dodys** [comment]
Looks good to me and the tests are passing can I move forward with merging or should I wait any other changes?

**Comment by jschmidb** [comment]
I have no more pending changes, so go ahead!

**Pull request merged**